### ISSUE TITLE
xStreamBufferReset() leads to a panic() (IDFGH-6040)

### ISSUE THREAD

**Issue by scwgoire** [espressif/esp-idf#7725]
## Environment

- Development Kit:      ESP32-DevKitC
- Kit version (for WroverKit/PicoKit/DevKitC): v3 (I guess)
- Module or chip used:  ESP32-WROOM-32U
- IDF version (run ``git describe --tags`` to find it): v4.3.1
- Build System:         idf.py
- Compiler version (run ``xtensa-esp32-elf-gcc --version`` to find it): xtensa-esp32-elf-gcc (crosstool-NG esp-2021r1) 8.4.0
- Operating System:     macOS
- Using an IDE?: No
- Power Supply:         USB

## Problem Description

xStreamBufferReset() leads to a panic().
Looks like `taskENTER_CRITICAL( &pxStreamBuffer->xStreamBufferMux )` tries to acquire a mutex which leads us to a panic() in `lock_acquire_generic()` because a recursive lock is requested and `xPortCanYield()` says the code is running in ISR, which it shouldn't be, see my code example.

I either missed the whole point of `xStreamBufferReset` or there is something really fishy there...

### Expected Behavior

Should reset the stream buffer

### Actual Behavior

panic()

### Steps/Code to reproduce

1. Create a new project using `idf.py create-project`
2. Type the following code in you main source file
```cpp
#include <stdio.h>
#include "freertos/FreeRTOS.h"
#include "freertos/task.h"
#include "freertos/stream_buffer.h"

void app_main(void)
{
    StreamBufferHandle_t mySB = xStreamBufferCreate(16, 1);
    xStreamBufferReset(mySB);
    printf("DONE!\n");
    while (1)
        vTaskDelay(10000 / portTICK_PERIOD_MS); // 10s
}
```
The printf call is never reached.
That simple!

## Debug Logs

```
ELF file SHA256: 552106fb46f1b21f

Rebooting...
ets Jun  8 2016 00:22:57

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:7208
load:0x40078000,len:14308
load:0x40080400,len:3716
0x40080400: _init at ??:?

entry 0x40080680
I (27) boot: ESP-IDF v4.3.1-dirty 2nd stage bootloader
I (27) boot: compile time 14:49:14
I (27) boot: chip revision: 1
I (30) boot_comm: chip revision: 1, min. bootloader chip revision: 0
I (38) boot.esp32: SPI Speed      : 40MHz
I (42) boot.esp32: SPI Mode       : DIO
I (47) boot.esp32: SPI Flash Size : 2MB
I (51) boot: Enabling RNG early entropy source...
I (57) boot: Partition Table:
I (60) boot: ## Label            Usage          Type ST Offset   Length
I (67) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (75) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (82) boot:  2 factory          factory app      00 00 00010000 00100000
I (90) boot: End of partition table
I (94) boot_comm: chip revision: 1, min. application chip revision: 0
I (101) esp_image: segment 0: paddr=00010020 vaddr=3f400020 size=06c9ch ( 27804) map
I (120) esp_image: segment 1: paddr=00016cc4 vaddr=3ffb0000 size=02928h ( 10536) load
I (124) esp_image: segment 2: paddr=000195f4 vaddr=40080000 size=06a24h ( 27172) load
I (138) esp_image: segment 3: paddr=00020020 vaddr=400d0020 size=134b0h ( 79024) map
I (167) esp_image: segment 4: paddr=000334d8 vaddr=40086a24 size=04338h ( 17208) load
I (174) esp_image: segment 5: paddr=00037818 vaddr=50000000 size=00010h (    16) load
I (180) boot: Loaded app from partition at offset 0x10000
I (180) boot: Disabling RNG early entropy source...
I (195) cpu_start: Pro cpu up.
I (195) cpu_start: Starting app cpu, entry point is 0x40081024
0x40081024: call_start_cpu1 at /Users/greg/src/esp-idf/components/esp_system/port/cpu_start.c:141

I (182) cpu_start: App cpu up.
I (210) cpu_start: Pro cpu start user code
I (210) cpu_start: cpu freq: 160000000
I (210) cpu_start: Application information:
I (214) cpu_start: Project name:     mytest
I (219) cpu_start: App version:      1
I (224) cpu_start: Compile time:     Oct 18 2021 14:49:08
I (230) cpu_start: ELF file SHA256:  552106fb46f1b21f...
I (236) cpu_start: ESP-IDF:          v4.3.1-dirty
I (241) heap_init: Initializing. RAM available for dynamic allocation:
I (248) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (254) heap_init: At 3FFB31E8 len 0002CE18 (179 KiB): DRAM
I (261) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (267) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (273) heap_init: At 4008AD5C len 000152A4 (84 KiB): IRAM
I (281) spi_flash: detected chip: generic
I (284) spi_flash: flash io: dio
W (288) spi_flash: Detected size(4096k) larger than the size in the binary image header(2048k). Using the size in the binary image header.
I (302) cpu_start: Starting scheduler on PRO CPU.
I (0) cpu_start: Starting scheduler on APP CPU.

abort() was called at PC 0x4008254f on core 0
0x4008254f: lock_acquire_generic at /Users/greg/src/esp-idf/components/newlib/locks.c:139

Backtrace:0x40084ba3:0x3ffb59e0 0x400852f5:0x3ffb5a00 0x40089db2:0x3ffb5a20 0x4008254f:0x3ffb5a90 0x40082699:0x3ffb5ac0 0x40082712:0x3ffb5ae0 0x400d98e9:0x3ffb5b10 0x400d57bd:0x3ffb5dd0 0x400d574c:0x3ffb5e20 0x40087d53:0x3ffb5e50 0x40087a30:0x3ffb5e70 0x400d4062:0x3ffb5e90 0x400e32fc:0x3ffb5eb0 0x40087b3d:0x3ffb5ed0
0x40084ba3: panic_abort at /Users/greg/src/esp-idf/components/esp_system/panic.c:368
0x400852f5: esp_system_abort at /Users/greg/src/esp-idf/components/esp_system/system_api.c:112
0x40089db2: abort at /Users/greg/src/esp-idf/components/newlib/abort.c:46
0x4008254f: lock_acquire_generic at /Users/greg/src/esp-idf/components/newlib/locks.c:139
0x40082699: _lock_acquire_recursive at /Users/greg/src/esp-idf/components/newlib/locks.c:167
0x40082712: __retarget_lock_acquire_recursive at /Users/greg/src/esp-idf/components/newlib/locks.c:323
0x400d98e9: _vfiprintf_r at /builds/idf/crosstool-NG/.build/HOST-x86_64-apple-darwin12/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/vfprintf.c:853 (discriminator 2)
0x400d57bd: fiprintf at /builds/idf/crosstool-NG/.build/HOST-x86_64-apple-darwin12/xtensa-esp32-elf/src/newlib/newlib/libc/stdio/fiprintf.c:48
0x400d574c: __assert_func at /builds/idf/crosstool-NG/.build/HOST-x86_64-apple-darwin12/xtensa-esp32-elf/src/newlib/newlib/libc/stdlib/assert.c:58 (discriminator 8)
0x40087d53: spinlock_release at /Users/greg/src/esp-idf/components/esp_hw_support/include/soc/spinlock.h:149
 (inlined by) vPortCPUReleaseMutex at /Users/greg/src/esp-idf/components/freertos/port/xtensa/include/freertos/portmacro.h:168
 (inlined by) vPortExitCritical at /Users/greg/src/esp-idf/components/freertos/port/xtensa/port.c:462
0x40087a30: xStreamBufferReset at /Users/greg/src/esp-idf/components/freertos/stream_buffer.c:443
0x400d4062: app_main at /Users/greg/src/mytest/main/mytest.c:9
0x400e32fc: main_task at /Users/greg/src/esp-idf/components/freertos/port/port_common.c:133 (discriminator 2)
0x40087b3d: vPortTaskWrapper at /Users/greg/src/esp-idf/components/freertos/port/xtensa/port.c:168```

# One more thing
Cheers :-)

**Comment by AxelLin** [comment]
v4.3 branch still does not include this fix.

**Comment by Alvin1Zhang** [comment]
@AxelLin Thanks for the attention, we have back ported this fix to 4.3, but still under internal reviewing, will sync up here once the fix is available on GitHub.
